### ISSUE TITLE
Except for workflow, node behaviour should always be a class property

### ISSUE THREAD

**Issue by liamhuber** [pyiron/pyiron_workflow#190]
Right now we offer three fundamental ways of constructing a new node (either a `Function` or a `Macro`):
1. By decorating a function (which becomes the `Function.node_function` or `Macro.graph_creator`)
2. By subclassing `Function` or `Macro` and defining `Function.node_function ` or `Macro.graph_creator`
3. By instantiating a `Function` or `Macro` and setting `Function.node_function ` or `Macro.graph_creator` _as an instance property_.

So for the first two, information about how the node works (e.g. inputs and their names, and (at a minimum) how many outputs are returned) are available _at the class level_. While for the third route, by definition these are only available post-instantiation.

I would like to get rid of route (3) entirely and make `Function` and `Macro` abstract, so that actual instantiated nodes _always_ come from a class created either by a decorator or a manual sub-class.

Concretely, I want to disallow:

```python
from pyiron_workflow import Workflow

def my_func(x):
    return x + 1

node = Workflow.create.Function(my_func)
```

In favour of the other two existing routes:

```python
from pyiron_workflow import Workflow

@Workflow.wrap_as.function_node()
def MyFunc(x):
    return x + 1

node = MyFunc()
```

or 

```python
from pyiron_workflow.function import Function

class MyFunc(Function):
    @staticmethod
    def node_function(x):
        return x + 1

node = MyFunc()
```

This will help get rid of some edge cases, [e.g. in storage](https://github.com/pyiron/pyiron_workflow/blob/edac929d644a7de283e47348127d4e61c7417138/pyiron_workflow/node.py#L178) where `h5io` has trouble reading instances of the base classes where these methods were provided at instantiation. 

It also streamlines other ideas, e.g. it will allow us to get type hints and scrape output labels at the class level; while this isn't critical now, being able to scrape info like type hints before we've instantiated things is exactly the type of tool we need for building node recommendation lists like in ironflow.

Bottom line: I think this removes a _little bit_ of flexibility for users, but gives us robustness and fewer edge cases to worry about. In the long run, I think it makes the platform stronger.

@samwaseda and @JNmpi, I'll push a PR on this and once it's ready I'll probably ping you (Sam), but it's more important to me that you give feedback on the idea than the execution.

**Comment by JNmpi** [comment]
@liamhuber, what about the following example:

      Workflow.create.Function(my_func)

with

      def Function(my_func):
             wrapped_func = Workflow.wrap_as.function_node(my_func)
             return wrapped_function()    # instantiate node object

This construct would behave like your other examples. Would such a solution be excluded in your proposal?

**Comment by JNmpi** [comment]
@liamhuber, thanks for your prompt reply. I am glad that the example worked. I agree also that using .__new__ would be the way to go for the actual implementation. 

With respect to the various node definitions. I could live with the requirement of having to explicitly define functions. To make workflows FAIR, it must be shareable and thus node definitions must be explicit. An exception may be temporary nodes that are dynamically (and reproducibly) generated by the workflow. Since their construction schema is part of the workflow it would be perfectly ok if they are not stored.

**Comment by liamhuber** [comment]
Alright, so this was actually super sneaky. It turns out the actual loading was failing for the implicit nodes both inside a workflow _and_ when they were stand-alone -- the catch is that when they were stand-alone the failure was not critical.

When we save and load with `h5io`, we're actually storing a reference to the module and class name, then re-instantiating a new object from there. In `pyiron_workflow` this gets leveraged like this:
- Instantiate a new node
- After instantiation see if there's a save file
- If so, use it (with `h5io/tinybase`) to instantiate a new object and update yourself like: `inst = h5io.read_hdf5(...); node.__setstaet__(inst.__getstate__)`

In the case of the declared node in my example above `__main__.Declared` finds the _decorated function_ and so importing it correctly returns a node class. In contrast, when we implicitly define the node by passing an existing function to a decorator, the function still exists, and the new class name does not match with the variable holding the node class, i.e. `Implicit.__name__ == "implicit"`.

Thus, we can safely re-instantiate a node (`Implicit()`), but when deserializing we import `__main__.implicit` -- which is a function! For pure nodes `some_node.__setstate__(some_function.__getstate__())` winds up being permissible (although obviously won't get us to the actual desired state), but when we deserialize a _composite_ its children really do become functions instead of nodes and we run into critical failures.

How to solve this? Of course we could mangle the name of the implicitly created class; this would mean we don't silently import its underlying function, but it doesn't actually help us get the class back because the class is _fundamentally not importable_. This would remain a problem whether we use `__new__` or directly leverage the decorators in order to dynamically create the class.

It also can't be solved at the `__get(set)state__` level, since these are what's used by tools like `h5io` to get the content; the class and module data itself gets scraped at the higher level of the storage tool.

We also can't prevent the behaviour, since users don't need to use a decorator like `@` (which would leave the resulting node importable), but can do like above and `Implicit = Workflow.wrap_as.single_value_node()(implicit)` (which leaves the resulting node unimportable).

So the only option I see is to do a check at save time to see whether `type(node) is getattr(importlib.import_module(node.__class__.__module__), node.__class__.__name__)`. When there is a mismatch here we could:
- Fail hard
- OR Issue a warning and fall back on `cloudpickle`

Failing hard should be pretty easy to implement, but having a fall-back involves some control at the level of the storage tool. This should be pretty easy in `tinybase.storage`, where @pmrv has already [provided a fallback](https://github.com/pyiron/pyiron_contrib/blob/9f855efe7ee3cee23010a2b8d9a33dc395552187/pyiron_contrib/tinybase/storage.py#L126) to pickle objects under a different condition, so we could pull a similar trick. I'm skeptical of trying to do something similar `h5io`, as it messes with the paradigm too much. Ultimately, this drives me in the direction of taking the [time and money](https://github.com/orgs/pyiron/discussions/214#discussioncomment-8362831) to get `tinybase.storage.GenericStorage` broken out and robustly depending on a combination of `__get(set)state__` and `cloudpickle`.

Meanwhile, my comments about the benefits of having `node_function/graph_creator` signatures guaranteed to be available on a `classmethod` all still stand, so regardless of the hiccups with storage I want to proceed with the `__new__` attack.

This issue breaks into two:
- Guarantee node behaviour is class property by exploiting `__new__`
- Put some more guard rails around storage so we fail early (or never) when saving and loading instances of totally dynamic classes

**Comment by liamhuber** [comment]
Today I fought with using `__new__` to dynamically create a child class of `Function` and return that instead. Unfortunately, this doesn't play well with the `HasPost` metaclass, and I didn't manage to get it working -- I couldn't get past the point of calling `__init__` on the parent `Function` class _after_ the entire `HasPost` process had called on the dynamic child class.

My current plan is to go back to having `Function` (and `Macro`) be plain-old abstract classes, such that the users either call the wrapper functions or just manually make a sub-class. 

This doesn't do anything to help with the conflict between implicit class definitions and storage; as long as we are dynamically creating new classes with the wrapper, there's always the chance the user invokes the wrapper manually such that the class name and importable variable name differ (e.g. `MyClass = Workflow.wrap_as.single_value_node()(some_function)`, where `MyClass.__name__` is `some_function`, but if we import `some_function` we're going to get the function itself and not the node class). We could test for this inside the wrapper, to make sure it's _always_ being used as a decorator -- or we can just delay the check until storage time. Either way, it's orthogonal to `Function` and `Macro` being abstract.

**Comment by pmrv** [comment]
So I'm not following this discussion or all of the developments at all, but just a short observation from your last comment.  Maybe this is ruled out by other design constraints that I'm missing because I don't have an overview.

It seems that the fundamental difficulty here are dynamic classes.  You introduce dynamic classes to enable
```python
MyNodeClass = Workflow.wrap_as.single_value_node("out")(some_function)
```
but do you really need a new class for each function?  Could you not have a normal class that takes the function as an input?  On first glance this might make the input/output channels more complicated, but I think you still make some magic such that 
```
MyNode = Workflow.wrap_as.single_value_node("out")(some_function)
node.input.foobar = ...
```
does some consistency checks such that `foobar` really is in `some_function`'s kwargs. If you `MyNode()` that could return a copy unless you give also a `label` in which case it would return itself unless the label changed.

Anyway, may be irrelevant, but it stood out to me just passing by from my mails.
